File: BackEnd/src/data/datasources/mongodb/models/ReminderModel.ts
/**
 * @file ReminderModel.ts
 * @description Define los nombres de colección y tipos de documento MongoDB para
 * Reminder y ReminderHistory. SIN lógica de mapeo.
 * @module Reminders
 * @layer Data
 */

import { ObjectId } from 'mongodb';
import type { ReminderType } from '../../../../domain/entities/Reminder.js';
import type { ReminderResult } from '../../../../domain/entities/ReminderHistory.js';

/** Nombre de la colección de recordatorios */
export const REMINDER_COLLECTION = 'reminders';

/** Nombre de la colección de historial de recordatorios */
export const REMINDER_HISTORY_COLLECTION = 'reminder_history';

/**
 * Tipo del documento de Reminder en MongoDB.
 */
export interface ReminderDocument {
  _id:           ObjectId;
  plantId:       ObjectId;
  userId:        ObjectId;
  type:          ReminderType;
  scheduledDate: Date;
  message:       string;
  isCompleted:   boolean;
  suspended:     boolean;
  attempts:      number;
  createdAt:     Date;
}

/**
 * Tipo del documento de ReminderHistory en MongoDB.
 */
export interface ReminderHistoryDocument {
  _id:            ObjectId;
  reminderId:     ObjectId;
  processedAt:    Date;
  result:         ReminderResult;
  details?:       string;
  idempotencyKey: string;
}

File: FrontEnd/lib/data/models/reminder_model.dart
/// @file reminder_model.dart
/// @description Modelo de serialización de recordatorio para la capa de datos.
/// Solo fromJson/toJson. SIN lógica de negocio ni mapeo a entidad.
/// La conversión ReminderModel ↔ Reminder la realiza ReminderMapper.
/// @module Reminders
/// @layer Data
library;

// ═══════════════════════════════════════════════════════════════════════════════
// REMINDER MODEL
// ═══════════════════════════════════════════════════════════════════════════════

/// Modelo de serialización de recordatorio. Refleja la estructura JSON de la API.
///
/// Regla estricta: NO contiene lógica de negocio, validaciones ni referencias
/// a entidades de dominio. Solo fromJson y toJson.
class ReminderModel {
  const ReminderModel({
    required this.id,
    required this.plantId,
    required this.userId,
    required this.type,
    required this.scheduledDate,
    required this.message,
    required this.isCompleted,
    required this.suspended,
    required this.attempts,
    required this.createdAt,
  });

  final String  id;
  final String  plantId;
  final String  userId;
  final String  type;
  final String  scheduledDate; // ISO 8601 string tal como llega del servidor.
  final String  message;
  final bool    isCompleted;
  final bool    suspended;
  final int     attempts;
  final String  createdAt;     // ISO 8601 string.

  // ─── Deserialización ─────────────────────────────────────────────────────────

  factory ReminderModel.fromJson(Map<String, dynamic> json) => ReminderModel(
    id:            json['_id']           as String? ?? json['id'] as String,
    plantId:       json['plantId']       as String,
    userId:        json['userId']        as String,
    type:          json['type']          as String,
    scheduledDate: json['scheduledDate'] as String,
    message:       json['message']       as String,
    isCompleted:   json['isCompleted']   as bool? ?? false,
    suspended:     json['suspended']     as bool? ?? false,
    attempts:      json['attempts']      as int?  ?? 0,
    createdAt:     json['createdAt']     as String,
  );

  // ─── Serialización ────────────────────────────────────────────────────────────

  Map<String, dynamic> toJson() => {
    'id':            id,
    'plantId':       plantId,
    'userId':        userId,
    'type':          type,
    'scheduledDate': scheduledDate,
    'message':       message,
    'isCompleted':   isCompleted,
    'suspended':     suspended,
    'attempts':      attempts,
    'createdAt':     createdAt,
  };
}
